### PR TITLE
Feature/pubsub splitprocessor

### PR DESCRIPTION
**Pull request by cristipogacean**
split event processor into several separate processes

### CODE REVIEW

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Microsoft.Azure.IIoT.Services.Processor.Events.csproj [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\api\src\Microsoft.Azure.IIoT.Api\src\Microsoft.Azure.IIoT.Api.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Diagnostics.AppInsights\src\Microsoft.Azure.IIoT.Diagnostics.AppInsights.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Auth.ActiveDirectory\src\Microsoft.Azure.IIoT.Auth.ActiveDirectory.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Hub.Processor\src\Microsoft.Azure.IIoT.Hub.Processor.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Messaging.SignalR\src\Microsoft.Azure.IIoT.Messaging.SignalR.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Messaging.EventHub\src\Microsoft.Azure.IIoT.Messaging.EventHub.csproj" />
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Microsoft.Azure.IIoT.Services.Processor.Events.csproj]: Is EventHub client still needed in the events processor?   I believe the others are, just double checking.

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Microsoft.Azure.IIoT.Services.Processor.Events.csproj [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\components\opc-ua\src\Microsoft.Azure.IIoT.OpcUa.Registry\src\Microsoft.Azure.IIoT.OpcUa.Registry.csproj" />
+    <ProjectReference Include="..\..\..\..\components\opc-ua\src\Microsoft.Azure.IIoT.OpcUa.Subscriber\src\Microsoft.Azure.IIoT.OpcUa.Subscriber.csproj" />
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Microsoft.Azure.IIoT.Services.Processor.Events.csproj]: Subscriber is likely not needed.   Registry might be, but check if it also can be removed.

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.Azure.IIoT.OpcUa.Graph.Tests")]
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Properties/AssemblyInfo.cs]: Wrong namespace for this assembly.   Probably copy and paste, --- just remove the entire AssemblyInfo.cs from all processors as there are no unit tests.

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Runtime/Config.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        /// <inheritdoc/>
+        public string OpcUaOnboardingServiceUrl => _ia.OpcUaOnboardingServiceUrl;
+        /// <inheritdoc/>
+        public string OpcUaOnboardingServiceResourceId => _ia.OpcUaOnboardingServiceResourceId;
+        /// <inheritdoc/>
+        public string EventHubConnString => _eh.EventHubConnString;
+        /// <inheritdoc/>
+        public string EventHubPath => _eh.EventHubPath;
+        /// <inheritdoc/>
+        //public string ConsumerGroup => _eh.ConsumerGroup;
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Runtime/Config.cs]: Just remove this commented out line.

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Runtime/Uptime.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.IIoT.Services.Processor.Telemetry.Runtime {
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Events/src/Runtime/Uptime.cs]: Namespace - use "Events".

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Cdm/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Cdm.csproj [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+     <GenerateDocumentationFile>true</GenerateDocumentationFile>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="3.1.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.CommandLine" Version="3.1.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="3.1.0" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\api\src\Microsoft.Azure.IIoT.Api\src\Microsoft.Azure.IIoT.Api.csproj" />
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Cdm/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Cdm.csproj]: I think the API project reference can be removed here.

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux.csproj [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\api\src\Microsoft.Azure.IIoT.Api\src\Microsoft.Azure.IIoT.Api.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Diagnostics.AppInsights\src\Microsoft.Azure.IIoT.Diagnostics.AppInsights.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Auth.ActiveDirectory\src\Microsoft.Azure.IIoT.Auth.ActiveDirectory.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Hub.Processor\src\Microsoft.Azure.IIoT.Hub.Processor.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Messaging.SignalR\src\Microsoft.Azure.IIoT.Messaging.SignalR.csproj" />
+    <ProjectReference Include="..\..\..\..\common\src\Microsoft.Azure.IIoT.Messaging.EventHub\src\Microsoft.Azure.IIoT.Messaging.EventHub.csproj" />
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux.csproj]: Might be able remove the Subscriber, the API reference and the EventHub project reference.

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.Azure.IIoT.OpcUa.Graph.Tests")]
```
Comment (marcschier) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Properties/AssemblyInfo.cs]: Remove

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Runtime/Config.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    /// </summary>
+    public class Config : DiagnosticsConfig, IEventProcessorConfig, 
+        IEventHubConsumerConfig, ISignalRServiceConfig {
+
+        /// <inheritdoc/>
+        public string EventHubConnString => _eh.EventHubConnString;
+        /// <inheritdoc/>
+        public string EventHubPath => _eh.EventHubPath;
+        /// <inheritdoc/>
+        public string ConsumerGroup => "telemetryux";
```
Comment (mregen) [inline services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Runtime/Config.cs]: same as above? should probably be a parm in config as its defined in the arm template?

File: common/src/Microsoft.Azure.IIoT.Cdm/src/Services/CdmMessageProcessor.cs [hunk 10]
```diff
@@ -193,26 +192,36 @@ public class CdmMessageProcessor : ICdmClient {
                 _logger.Warning("Failed to send processed CDM data after {elapsed} : {message}",
                      sw.Elapsed, errorMessage);
             }
-            finally {
-                _lock.Release();
+            finally {                
+                _cacheList.Clear();
```
Comment (mregen) [inline common/src/Microsoft.Azure.IIoT.Cdm/src/Services/CdmMessageProcessor.cs]: just being super safe: _cacheList?.Clear...

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Runtime/Config.cs (hunk 9 above)
Comment (marcschier) [reply services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Ux/src/Runtime/Config.cs]: In this case they are hardcoded, they always connect to the same consumer group.   You just have to make sure such group is defined.

File: services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Cdm/src/Runtime/Config.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    public class Config : DiagnosticsConfig, IEventProcessorConfig,
+        IEventHubConsumerConfig, ICdmClientConfig {
+
+        /// <inheritdoc/>
+        public string EventHubConnString => _eh.EventHubConnString;
+        /// <inheritdoc/>
+        public string EventHubPath => _eh.EventHubPath;
+        /// <inheritdoc/>
+        //public string ConsumerGroup => _eh.ConsumerGroup;
+        public string ConsumerGroup => "telemetrycdm";
```
[the remark this answers is not in the document]
Comment (cristipogacean) [reply services/src/Microsoft.Azure.IIoT.Services.Processor.Telemetry.Cdm/src/Runtime/Config.cs]: In this case they are hardcoded, they always connect to the same consumer group.  

---
In reply to: [366757320](https://github.com/Azure/Industrial-IoT/pull/195#discussion_r366757320) [](ancestors = 366757320)

**Pull request merged**